tldw_Server_API/app/core/StudyPacks/generation_service.py

```
@@ -0,0 +1,504 @@
+        locator = self._merge_citation_locator(bundle_item, raw_citation.get("locator"))
+        return {
+            "source_type": bundle_item.source_type,
+            "source_id": bundle_item.source_id,
+            "citation_text": citation_text,
+            "locator": locator,
+        }
+
+    def _merge_citation_locator(
+        self,
+        bundle_item: StudySourceBundleItem,
+        raw_locator: Any,
```

Ruff (0.15.7)

[warning] Dynamically typed expressions (typing.Any) are disallowed in `raw_locator`

(ANN401)